keps/sig-node/20200327-not-inject-kubelet-to-its-components.md

```diff
@@ -0,0 +1,318 @@
+    "k8s.io/kubernetes/pkg/kubelet/apis"
+    "k8s.io/kubernetes/pkg/kubelet/config"
+    "k8s.io/kubernetes/pkg/util/node"
+    "k8s.io/kubernetes/pkg/util/taints"
+    "k8s.io/kubernetes/pkg/volume/util"
+)
+
+// BasicInfo's members are migrated from Kubelet.
+type BasicInfo struct {
+
+    Hostname string
+
+    NodeName types.NodeName
+
+    Clock clock.Clock
+
+    RegisterNode bool
+
+    RegistrationCompleted bool
+
+    RootDirectory string
+
+    KubeClient clientset.Interface
+
+    MasterServiceNamespace string
+
+    nodeLabels map[string]string
+
+    Recorder record.EventRecorder
+
+    ServiceLister serviceLister
+
+    NodeLister corelisters.NodeLister
+
+    providerID string
+
+    externalCloudProvider bool
+
+    Cloud cloudprovider.Interface
+
+    registerSchedulable bool
+
+    registerWithTaints []api.Taint
+
+    enableControllerAttachDetach bool
+
+    ExperimentalHostUserNamespaceDefaulting bool
+
+    NodeStatusFuncs []func(*v1.Node) error
+
+    KeepTerminatedPodVolumes bool // DEPRECATED
+}
+
+// these functions are kubelet's method now.
+type Interface struct {
+    // current kubelet's getHostIPAnyWay() function.
+    GetHostIPAnyWay() (net.IP, error)
+
+    // current kubelet's getNodeAnyWay function.
+    GetNodeAnyWay() (*v1.Node, error)
+
+    // current kubelet's GetNode function.
+    GetNode()
+
+    // current kubelet's initialNode function.
+    InitialNode(ctx context.Context) (*v1.Node, error)
+
+    // current kubelet.setNodeStatus.
+    SetNodeStatus(node *v1.Node)
+
+    // following get dir methods locate in kubelet_getters.go now belongs to Kubelet currently.
+    GetRootDir() string
+    SetRootDir(dir string)
+    GetPodsDir() string
+    GetPluginsDir() string
+    GetPluginsRegistrationDir() string
+    GetPluginDir(pluginName string) string
+    GetVolumeDevicePluginsDir() string
+    GetVolumeDevicePluginDir(pluginName string)
+    GetPodDir(podUID types.UID) string
+    GetPodVolumeSubpathsDir(podUID types.UID) string
+    GetPodVolumesDir(podUID types.UID) string
+    GetPodVolumeDir(podUID types.UID, pluginName string, volumeName string) string
+    GetPodVolumeDevicesDir(podUID types.UID) string
+    GetPodVolumeDeviceDir(podUID types.UID, pluginName string) string
+    GetPodPluginsDir(podUID types.UID) string
+    GetPodPluginDir(podUID types.UID, pluginName string) string
+    GetPodContainerDir(podUID types.UID, ctrName string) string
+    GetPodResourcesDir() string
+}
+```
+Inject `Kubelet.BasicInfo` to Kubele's component and remove these members from `Kubelet`.
+
+### Init GenericRuntimeManager and ResourceAnalyzer elaborately
+
+For `kubeGenericRuntimeManager` initialization.
+Current kubeGenericRuntimeManager use `kubelet` as `runtimeHelper` and `podStateProvider` now.
+
+Changes of `kubeGenericRuntimeManager` and `containerGC`:
+```go
+type kubeGenericRuntimeManager struct {
+    ...
+
+    // Remove.
+    runtimeHelper kubecontainer.RuntimeHelper
```

Comment:
This seems like it is getting more complex, not less.  Am I understanding the proposed changes correctly?

Reply:
current code makes `Kubelet` struct as runtimeHelper, IMO we should put specific dependencies rather than the whole `Kubelet`.

Reply:
It doesn't really matter that we are putting the _whole_ kubelet somewhere as a dependency.  What matters is the size of the interface, as those are the methods actually offered to the dependent package.

In other words, if we want to simplify the dependent package, we mostly care about simplifying the interface it depends on.  Splitting up the kubelet into multiple parts won't make the dependent packages much simpler.  The reason we would want to split the kubelet up into parts is to simplify the kubelet, and make it easier to understand.

Reply:
Sorry, I mean put Kubelet as a helper interface caused methods all implemented in Kubelet struct, it makes kubelet root directory a mess. For example, `GeneratePodHostNameAndDomain`, `GetPodCgroupParent`, `GenerateRunContainerOptions`, `makeBlockVolumes`, etc (and their sub functions) all belong to Kubelet, however they are only used by `genericRuntimeManager` or `fsResourceAnalyzer`. Shall we converge them to its sub components rather than whole Kubelet?

Reply:
> The reason we would want to split the kubelet up into parts is to simplify the kubelet, and make it easier to understand.

Agree with you. I think it's a pre work, maybe I would split kubelet better rather than put meta info to `BasicInfo` now.